File: sdk/npm/src/providers/apihz.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";
import { getConfig } from "../config";

const CHANNEL: Channel = "apihz";
const BASE_URL = "https://cn.apihz.cn";
const TOK_PREFIX = "apihz1:";

/** apihz 官方公共账号（共享频次），未配置自有 id/key 时使用 */
const PUBLIC_ID = "88888888";
const PUBLIC_KEY = "88888888";

const UA =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36";

/** 可用收信域名（apihz 自研，MX 指向 mail.apimail.* 自建服务器） */
const DOMAINS = ["apimail.email", "apimail.vip"];

/** 读取 apihz 调用凭据：优先配置/环境变量，回退官方公共账号 */
function getCredentials(): { id: string; key: string } {
  const cfg = getConfig();
  return {
    id: cfg.apihzId || PUBLIC_ID,
    key: cfg.apihzKey || PUBLIC_KEY,
  };
}

function randomLocal(n: number): string {
  const chars = "abcdefghijklmnopqrstuvwxyz0123456789";
  let r = "";
  for (let i = 0; i < n; i++) {
    r += chars[Math.floor(Math.random() * chars.length)];
  }
  return r;
}

/** 生成 8-20 位随机密码（读信必须携带创建时的密码） */
function randomPassword(): string {
  const chars =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  let r = "";
  for (let i = 0; i < 12; i++) {
    r += chars[Math.floor(Math.random() * chars.length)];
  }
  return r;
}

/** token 编码：前缀 + base64(JSON{mail,pwd})，读信时解出 pwd */
function encToken(mail: string, pwd: string): string {
  const raw = JSON.stringify({ mail, pwd });
  return TOK_PREFIX + Buffer.from(raw, "utf8").toString("base64");
}

function decToken(tok: string): { mail: string; pwd: string } {
  if (!tok.startsWith(TOK_PREFIX)) {
    throw new Error("apihz: 无效 token");
  }
  let raw: string;
  try {
    raw = Buffer.from(tok.slice(TOK_PREFIX.length), "base64").toString("utf8");
  } catch {
    throw new Error("apihz: 无效 token");
  }
  const o = JSON.parse(raw) as { mail?: string; pwd?: string };
  const mail = (o.mail ?? "").trim();
  const pwd = (o.pwd ?? "").trim();
  if (!mail || !pwd) throw new Error("apihz: 无效 token");
  return { mail, pwd };
}

/**
 * 创建 apihz（接口盒子）临时邮箱
 * GET /api/mail/mailcache.php?id=&key=&domain=&name=&pwd=&buytype=0
 * 有效期 10 分钟，读信必须携带创建时的 pwd
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const { id, key } = getCredentials();
  const domain = DOMAINS[Math.floor(Math.random() * DOMAINS.length)];
  const name = randomLocal(10);
  const pwd = randomPassword();

  const url =
    `${BASE_URL}/api/mail/mailcache.php?id=${encodeURIComponent(id)}&key=${encodeURIComponent(key)}` +
    `&domain=${encodeURIComponent(domain)}&name=${encodeURIComponent(name)}&pwd=${encodeURIComponent(pwd)}&buytype=0`;

  const resp = await fetchWithTimeout(url, {
    method: "GET",
    headers: { "User-Agent": UA, Accept: "application/json" },
  });

  if (!resp.ok) {
    throw new Error(`apihz: 创建邮箱失败 ${resp.status}`);
  }

  const data = (await resp.json()) as ApihzCreateResponse;
  if (data.code !== 200 || !data.mail) {
    throw new Error(`apihz: 创建邮箱失败 ${data.msg || JSON.stringify(data)}`);
  }

  /* 优先使用响应回传的 pwd（与请求一致），确保读信密码正确 */
  const finalPwd = data.pwd || pwd;

  return {
    channel: CHANNEL,
    email: data.mail,
    token: encToken(data.mail, finalPwd),
  };
}

/**
 * 获取 apihz 邮件列表
 * GET /api/mail/mailgetlist.php?id=&key=&mail=&pwd=&page=1
 */
export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  if (!token) {
    throw new Error("apihz: 缺少 token");
  }
  const { mail, pwd } = decToken(token);
  const { id, key } = getCredentials();
  const addr = mail || email;

  const url =
    `${BASE_URL}/api/mail/mailgetlist.php?id=${encodeURIComponent(id)}&key=${encodeURIComponent(key)}` +
    `&mail=${encodeURIComponent(addr)}&pwd=${encodeURIComponent(pwd)}&page=1`;

  const resp = await fetchWithTimeout(url, {
    method: "GET",
    headers: { "User-Agent": UA, Accept: "application/json" },
  });

  if (!resp.ok) {
    throw new Error(`apihz: 获取邮件失败 ${resp.status}`);
  }

  const data = (await resp.json()) as ApihzListResponse;
  if (data.code !== 200 || !Array.isArray(data.data)) {
    return [];
  }

  return data.data.map((msg, idx) =>
    normalizeEmail(
      {
        id: String(msg.time1 ?? idx),
        from: msg.frommail || msg.fromname || "",
        to: addr,
        subject: msg.subject || "",
        text: msg.content || "",
        html: msg.content || "",
        date: msg.time2 || (msg.time1 ? String(msg.time1) : ""),
        isRead: false,
        attachments: [],
      },
      addr,
    ),
  );
}

type ApihzCreateResponse = {
  code?: number;
  msg?: string;
  mail?: string;
  pwd?: string;
  endtime?: string;
};

type ApihzMessage = {
  frommail?: string;
  fromname?: string;
  subject?: string;
  time1?: number;
  time2?: string;
  content?: string;
};

type ApihzListResponse = {
  code?: number;
  msg?: string;
  num?: string | number;
  data?: ApihzMessage[];
};

File: sdk/go/provider/apihz.go
package provider

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

// apihz（接口盒子）：需 id/key 凭据的临时邮箱，邮箱有效期 10 分钟。
// 读信须携带创建时的 pwd，故 token 同时保存 mail + pwd。

const (
	apihzBaseURL   = "https://cn.apihz.cn"
	apihzTokPrefix = "apihz1:"
	// apihz 官方公共账号（共享频次），未配置自有 id/key 时使用
	apihzPublicID  = "88888888"
	apihzPublicKey = "88888888"
	apihzUA        = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36"
)

// apihzDomains 可用收信域名（apihz 自研，MX 指向 mail.apimail.* 自建服务器）
var apihzDomains = []string{"apimail.email", "apimail.vip"}

// apihzCredentials 读取 apihz 调用凭据：优先配置/环境变量，回退官方公共账号
func apihzCredentials() (id, key string) {
	id, key = apihzPublicID, apihzPublicKey
	if GetConfigSnapshot != nil {
		cfg := GetConfigSnapshot()
		if s := strings.TrimSpace(cfg.ApihzID); s != "" {
			id = s
		}
		if s := strings.TrimSpace(cfg.ApihzKey); s != "" {
			key = s
		}
	}
	return id, key
}

func apihzRandomLocal(length int) string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	buf := make([]byte, length)
	for i := range buf {
		buf[i] = chars[rand.Intn(len(chars))]
	}
	return string(buf)
}

// apihzRandomPassword 生成 12 位随机密码（读信必须携带创建时的密码）
func apihzRandomPassword() string {
	const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
	buf := make([]byte, 12)
	for i := range buf {
		buf[i] = chars[rand.Intn(len(chars))]
	}
	return string(buf)
}

// apihzEncodeToken token 编码：前缀 + base64(JSON{mail,pwd})，读信时解出 pwd
func apihzEncodeToken(mail, pwd string) string {
	raw, _ := json.Marshal(struct {
		Mail string `json:"mail"`
		Pwd  string `json:"pwd"`
	}{Mail: mail, Pwd: pwd})
	return apihzTokPrefix + base64.StdEncoding.EncodeToString(raw)
}

// apihzDecodeToken 去前缀 → base64 解码 → JSON 解析，取出 mail/pwd
func apihzDecodeToken(token string) (mail, pwd string, err error) {
	if !strings.HasPrefix(token, apihzTokPrefix) {
		return "", "", fmt.Errorf("apihz: 无效 token")
	}
	raw, err := base64.StdEncoding.DecodeString(token[len(apihzTokPrefix):])
	if err != nil {
		return "", "", fmt.Errorf("apihz: 无效 token")
	}
	var o struct {
		Mail string `json:"mail"`
		Pwd  string `json:"pwd"`
	}
	if err := json.Unmarshal(raw, &o); err != nil {
		return "", "", fmt.Errorf("apihz: 无效 token")
	}
	mail = strings.TrimSpace(o.Mail)
	pwd = strings.TrimSpace(o.Pwd)
	if mail == "" || pwd == "" {
		return "", "", fmt.Errorf("apihz: 无效 token")
	}
	return mail, pwd, nil
}

type apihzCreateResponse struct {
	Code    int    `json:"code"`
	Msg     string `json:"msg"`
	Mail    string `json:"mail"`
	Pwd     string `json:"pwd"`
	Endtime string `json:"endtime"`
}

type apihzMessage struct {
	Frommail string      `json:"frommail"`
	Fromname string      `json:"fromname"`
	Subject  string      `json:"subject"`
	Time1    json.Number `json:"time1"`
	Time2    string      `json:"time2"`
	Content  string      `json:"content"`
}

type apihzListResponse struct {
	Code int             `json:"code"`
	Msg  string          `json:"msg"`
	Num  json.RawMessage `json:"num"`
	Data []apihzMessage  `json:"data"`
}

// ApihzGenerate 创建 apihz（接口盒子）临时邮箱
// GET /api/mail/mailcache.php?id=&key=&domain=&name=&pwd=&buytype=0
// 有效期 10 分钟，读信必须携带创建时的 pwd
func ApihzGenerate() (*CreatedMailbox, error) {
	id, key := apihzCredentials()
	domain := apihzDomains[rand.Intn(len(apihzDomains))]
	name := apihzRandomLocal(10)
	pwd := apihzRandomPassword()

	u := fmt.Sprintf("%s/api/mail/mailcache.php?id=%s&key=%s&domain=%s&name=%s&pwd=%s&buytype=0",
		apihzBaseURL, url.QueryEscape(id), url.QueryEscape(key),
		url.QueryEscape(domain), url.QueryEscape(name), url.QueryEscape(pwd))

	req, err := http.NewRequest("GET", u, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", apihzUA)
	req.Header.Set("Accept", "application/json")

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := CheckHTTPStatus(resp, "apihz generate"); err != nil {
		return nil, err
	}

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var data apihzCreateResponse
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil, fmt.Errorf("apihz: 创建邮箱返回非 JSON: %s", string(raw[:min(len(raw), 120)]))
	}
	if data.Code != 200 || data.Mail == "" {
		msg := data.Msg
		if msg == "" {
			msg = string(raw)
		}
		return nil, fmt.Errorf("apihz: 创建邮箱失败 %s", msg)
	}

	// 优先使用响应回传的 pwd（与请求一致），确保读信密码正确
	finalPwd := data.Pwd
	if finalPwd == "" {
		finalPwd = pwd
	}

	return &CreatedMailbox{
		Channel: "apihz",
		Email:   data.Mail,
		Token:   apihzEncodeToken(data.Mail, finalPwd),
	}, nil
}

// ApihzGetEmails 获取 apihz 邮件列表
// GET /api/mail/mailgetlist.php?id=&key=&mail=&pwd=&page=1
// token 已含 mail 与 pwd，读信只需 token
func ApihzGetEmails(token string) ([]NormEmail, error) {
	if token == "" {
		return nil, fmt.Errorf("apihz: 缺少 token")
	}
	mail, pwd, err := apihzDecodeToken(token)
	if err != nil {
		return nil, err
	}
	id, key := apihzCredentials()

	u := fmt.Sprintf("%s/api/mail/mailgetlist.php?id=%s&key=%s&mail=%s&pwd=%s&page=1",
		apihzBaseURL, url.QueryEscape(id), url.QueryEscape(key),
		url.QueryEscape(mail), url.QueryEscape(pwd))

	req, err := http.NewRequest("GET", u, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", apihzUA)
	req.Header.Set("Accept", "application/json")

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := CheckHTTPStatus(resp, "apihz getEmails"); err != nil {
		return nil, err
	}

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var data apihzListResponse
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil, err
	}
	if data.Code != 200 || len(data.Data) == 0 {
		return []NormEmail{}, nil
	}

	emails := make([]NormEmail, 0, len(data.Data))
	for _, msg := range data.Data {
		from := msg.Frommail
		if from == "" {
			from = msg.Fromname
		}
		date := msg.Time2
		if date == "" {
			date = msg.Time1.String()
		}
		flat := map[string]interface{}{
			"id":      msg.Time1.String(),
			"from":    from,
			"to":      mail,
			"subject": msg.Subject,
			"text":    msg.Content,
			"html":    msg.Content,
			"date":    date,
			"isRead":  false,
		}
		emails = append(emails, NormalizeMap(flat, mail))
	}
	return emails, nil
}
